### PR TITLE
Update to latest rust nightly (2021-01-29)

### PR DESCRIPTION
**Pull request by KronicDeth**

### CODE REVIEW

File: .github/workflows/wasm32-unknown-unknown-macOS.yml [hunk 1]
```diff
@@ -11,15 +11,15 @@ jobs:
...
+          toolchain: nightly-2021-01-29
           override: true
       - name: Cache Cargo
         uses: actions/cache@v2
         with:
           path: |
             ~/.cargo/registry
             ~/.cargo/git
-          key: ${{ github.workflow }}-${{ github.job }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          key: ${{ github.workflow }}-${{ github.job }}-toolchain-nightly-2021-01-29-cargo-${{ hashFiles('**/Cargo.lock') }}
```
Comment (bitwalker) [inline .github/workflows/wasm32-unknown-unknown-macOS.yml:22]: You can replace the hardcoding of the toolchain name in the key with `${{ env.INPUT_TOOLCHAIN }}`

File: compiler/codegen/src/lib.rs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#![deny(warnings)]
```
Comment (bitwalker) [inline compiler/codegen/src/lib.rs:1]: As a rule with the amount of churn in the codegen/compiler crates, I'd prefer to not have this set - it makes it far far too annoying to iterate during development

File: compiler/codegen_llvm/lib/lumen/EIR/IR/EIROps.cpp [hunk 3]
```diff
@@ -2414,8 +2414,10 @@ static LogicalResult verify(MallocOp op) {
         if (!termType.isBoxable())
             return op.emitOpError("cannot malloc an unboxable term type");
 
-        Optional<Value> arityVal = op.arity();
-        if (arityVal.hasValue()) {
+        // Optional<arity> in `arguments` is `Value()`, that is `Value(nullptr)` if not given and not an
```
Comment (bitwalker) [inline compiler/codegen_llvm/lib/lumen/EIR/IR/EIROps.cpp:2417]: Can we revert any changes to the C++ code in this PR? I've made way too many changes locally for me to be able to merge them cleanly after this. A massive amount has changed. As an alternative, you could open a separate PR that contains these changes, and I can manually apply any changes that still carry over, but there will be no way for me to sanely merge against this.

**Review by bitwalker — CHANGES_REQUESTED** [review]
Looks good overall, just one change I'd like to request about the C++ code, and other than that a couple suggestions.

File: compiler/codegen/src/lib.rs (hunk 2 above)
Comment (KronicDeth) [reply compiler/codegen/src/lib.rs:1]: I can turn it back off, I turned it on to fix the old warnings, so I could see the new warnings I made with the updates easier.

File: compiler/codegen_llvm/lib/lumen/EIR/IR/EIROps.cpp (hunk 3 above)
Comment (KronicDeth) [reply compiler/codegen_llvm/lib/lumen/EIR/IR/EIROps.cpp:2417]: Not this one, but the one below about the `op.size()` is required to get the `lumen` compiler tests to pass at the same level as develop.  With @hansihe's changes to EIR to fix the binary code to not push a single byte at a time, but the whole binary, line 2747-2748 triggered an `assert`.

If I remove this changes the `compiler` workflows will be broken once this merges.

**Review by bitwalker — APPROVED** [review]

**Pull request merged**